### ISSUE TITLE
PHP INI setting, include_path - Restriction

### ISSUE THREAD

**Issue by uniflare** [koraktor/steam-condenser#58]
A lot of webhosts seem to restrict modifying PHP INI Variables even temporarily, I would reccommend using a require/include wrapper function to handle including files etc;

```
function c_require_once ($file){
   // CONSTANT PATH TO LIB: LIB_PATH
   if(!file_exists(LIB_PATH.$file)){
      trigger new FileExistException;
      return FALSE;
   }else{
      require_once(LIB_PATH.$file);
      return TRUE;
   }
}
```

Would make your script a little more portable, with minimal modification (open every php file in notepad++, replace "require_once" with "c_require_once" in all files, would take less than 1 minute :P).

-cb-

**Comment by koraktor** [comment]
I'm not really into the restrictions some hosts are using... might `set_include_path()` be an option?

**Comment by koraktor** [comment]
Ok, seems like this won't work either, if there are active path restrictions. So it seems like we'll need some workaround. Maybe I'll have to revive `__autoload()`.

**Comment by uniflare** [comment]
You can use    set_include_path(). But even this - some free hosts (admittedly much less) will still restrict this, and even some paid hosts will not let you change the include_path regardless (although i cannot see any security implications by using this function). I would imagine most people would try to find a new free host, in which case they will google free host and should come up with a suitable provider.

Also, using the include_path to deal with mapping your includes  _can_ be troublesome, but since your only using a single include_path it should be OK. Until you try to use a filename that already exists in someones own include_path setting.

Conclusion:
    set_include_path().
Seems a reasonable alternative, although using dynamic includes can generally lead to file consistency problems (although as mentioned a single include path is usually OK).

Regards,
-cb-

**Comment by koraktor** [comment]
So... another alternative would be using `dirname(__FILE__)` to have relative include paths, like `require_once dirname(__FILE__) . '/../../exceptions/PacketFormatException.php';`

PS: I'm not using `__DIR__` here because it has been initially introduced in PHP 5.3.

**Comment by uniflare** [comment]
PHP is very good with relative paths so technically you don't need the absolute path that dirname gives.

The best option i can think of, and easiest, is either - my original solution, or your sub solution of set_include_path();

Though the most compatible would be to use a CONSTANT to hold the root of the index file (though it's not needed), and to just use relative paths;
     require_once("../lib/InetAddress.php");
     require_once(INDEX_ROOT."/lib/InetAddress.php");

**Comment by koraktor** [comment]
To be honest I can't get relative paths to work because PHP will use the path of the currently _executed_ file as the base path, not the path of the currently _included_ file.
`dirname(__FILE__)` works and introducing a constant to replace this really seems strange to me.

**Comment by uniflare** [comment]
You mean you want to stick to OOP logic and make these classes completely independant by not having to know the directory structure of this specific project?

I don't get why you would prefer dirname as it complicates the issue of back-referencing directories. Where as using general relative paths is much easier to understand and implement. For example:

Your file structure is so that you need to know the directories or "locations" of each file.
    root
     - index.php
     - lib
     - - InetAddress.php
     - - steam
     - - - SteamPlayer.php

As you can see, to include the file "InetAddress" form inside SteamPlayer (or any other sub-directory file), using dirname() would complicated the issue:

```
include(dirname(__FILE__)."../InetAddress.php"); // Back reference to an unknown directory
```

vs.
    include("./lib/steam/SteamPlayer.php"); // Full Relative path to file, easily understood/found and maintained.

The _executed_ file should always be the index file, the controller, so you should have no problems. (index file would contain for ex, a Switch of Page Indexes, which would then include the necessary file to provide the output requested - with defaults of course).

-cb-

**Comment by uniflare** [comment]
More Info:

Using a constant is one way to define the current lib/classes directory, this way you can keep your OOP standards;

```
include(LIB_PATH."/steam/SteamPlayer.php"); // Much more portable class code, of course these classes need/depend on each other so this is as Portable as classes get with PHP.
```

-cb-

**Comment by koraktor** [comment]
The problem is that the "relative" way of including files only works as long as someone has the library in his `lib` subdirectory. As soon as people would use something like `lib/steam-condenser` or `deps` or whatever this will stop working.

In my opinion this is a shortcoming of PHP in general.
Everyone thinks of PHP as a website generator. Ok, it has been in the first place and it is used like that most of the time. Nevertheless it is a full-featured programming language - or at least it should be. Things like this really annoy programmers used to other languages.

But back to the solution of the problem:
I might have a good and clean solution that should work for every case. A new file, maybe `steam-condenser.php` (much like Ruby's [steam-condenser.rb](http://github.com/koraktor/steam-condenser/blob/master/ruby/lib/steam-condenser.rb), should define the library path in a constant (`STEAM_CONDENSER_PATH` to avoid conflicts) and include some of the main classes like `SourceServer`, `GoldSrcServer` and `SteamId`.

I think that should work fine and still be "usage agnostic".

**Comment by koraktor** [comment]
So... what do you think about this [proposal](http://github.com/koraktor/steam-condenser/compare/master...issue-58)?

**Comment by uniflare** [comment]
I think you have hit the nail on the head so to speak. :P.
It's been great communicating with you :).

-Chemical Bliss

(if you want help or ideas/advice at any point let me know by sending a PM and ill give you my contact details.)

**Comment by koraktor** [comment]
PHP: Improved loading

Closed by 20163d50defefe734d1e128301cff45a809076d3
